File: trunk/user/mtr/mtr-0.92/test/cmdparse.py
#!/usr/bin/env python
#
#   mtr  --  a network diagnostic tool
#   Copyright (C) 2016  Matt Kimball
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License version 2 as
#   published by the Free Software Foundation.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#

'''Test mtr-packet's command parsing.'''


import time
import unittest

import mtrpacket


class TestCommandParse(mtrpacket.MtrPacketTest):
    '''Test cases with malformed commands and version checks'''

    def test_unknown_command(self):
        'Test sending a command unknown to mtr-packet'

        self.write_command('13 argle-bargle')
        self.assertEqual(self.read_reply(), '13 unknown-command')

    def test_malformed_command(self):
        'Test sending a malformed command request to mtr-packet'

        self.write_command('malformed')
        self.assertEqual(self.read_reply(), '0 command-parse-error')

    def test_exit_on_stdin_closed(self):
        '''Test that the packet process terminates after stdin is closed

        Test that, when outstanding requests are complete, the process
        terminates following stdin being closed.'''

        self.write_command('15 send-probe ip-4 8.8.254.254 timeout 1')
        self.packet_process.stdin.close()
        time.sleep(2)
        self.read_reply()
        exit_code = self.packet_process.poll()
        self.assertIsNotNone(exit_code)

    def test_invalid_argument(self):
        'Test sending invalid arguments with probe requests'

        bad_commands = [
            '22 send-probe',
            '23 send-probe ip-4 str-value',
            '24 send-probe ip-4 8.8.8.8 timeout str-value',
            '25 send-probe ip-4 8.8.8.8 ttl str-value',
        ]

        for cmd in bad_commands:
            self.write_command(cmd)
            reply = self.parse_reply()
            self.assertEqual(reply.command_name, 'invalid-argument')

    def test_versioning(self):
        'Test version checks and feature support checks'

        feature_tests = [
            ('31 check-support feature ip-4', 'ok'),
            ('32 check-support feature send-probe', 'ok'),
            ('33 check-support feature bogus-feature', 'no')
        ]

        self.write_command('30 check-support feature version')
        reply = self.parse_reply()
        self.assertEqual(reply.token, 30)
        self.assertEqual(reply.command_name, 'feature-support')
        self.assertIn('support', reply.argument)

        for (request, expected) in feature_tests:
            self.write_command(request)
            reply = self.parse_reply()
            self.assertEqual(reply.command_name, 'feature-support')
            self.assertIn('support', reply.argument)
            self.assertEqual(reply.argument['support'], expected)

    def test_command_overflow(self):
        'Test overflowing the incoming command buffer'

        big_buffer = 'x' * (64 * 1024)
        self.write_command(big_buffer)

        reply = self.read_reply()
        self.assertEqual(reply, '0 command-buffer-overflow')


if __name__ == '__main__':
    mtrpacket.check_running_as_root()
    unittest.main()

File: trunk/user/mtr/mtr-0.92/packet/cmdparse.c
/*
    mtr  --  a network diagnostic tool
    Copyright (C) 2016  Matt Kimball

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "cmdparse.h"

#include <ctype.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>

/*
    NUL terminate the whitespace separated tokens in the command string.
    This modifies command_string in-place with NUL characters.
    Fill the tokens array with pointers to the tokens, and return the
    number of tokens found.
*/
static
int tokenize_command(
    char **tokens,
    int max_tokens,
    char *command_string)
{
    int token_count = 0;
    int on_space = 1;
    int i;

    for (i = 0; command_string[i]; i++) {
        if (on_space) {
            if (!isspace((unsigned char) command_string[i])) {
                /*  Take care not to exceed the token array length  */
                if (token_count >= max_tokens) {
                    return -1;
                }

                tokens[token_count++] = &command_string[i];
                on_space = 0;
            }
        } else {
            if (isspace((unsigned char) command_string[i])) {
                command_string[i] = 0;
                on_space = 1;
            }
        }
    }

    return token_count;
}

/*
    Parse a command string (or command reply string) into a command_t
    structure for later semantic interpretation.  Returns EINVAL if the
    command string is unparseable or zero for success.

    comamnd_string will be modified in-place with NUL characters terminating
    tokens, and the command_t will use pointers to the conents of
    command_string without copying, so any interpretation of the
    command_t structure requires that the command_string memory has not yet
    been freed or otherwise reused.
*/
int parse_command(
    struct command_t *command,
    char *command_string)
{
    char *tokens[MAX_COMMAND_TOKENS];
    int token_count;
    int i;

    memset(command, 0, sizeof(struct command_t));

    /*  Tokenize the string using whitespace  */
    token_count =
        tokenize_command(tokens, MAX_COMMAND_TOKENS, command_string);
    if (token_count < 2) {
        errno = EINVAL;
        return -1;
    }

    /*  Expect the command token to be a numerical value  */
    errno = 0;
    command->token = strtol(tokens[0], NULL, 10);
    if (errno) {
        errno = EINVAL;
        return -1;
    }
    command->command_name = tokens[1];

    /*
       The tokens beyond the command name are expected to be in
       name, value pairs.
     */
    i = 2;
    command->argument_count = 0;
    while (i < token_count) {
        /*  It's an error if we get a name without a key  */
        if (i + 1 >= token_count) {
            errno = EINVAL;
            return -1;
        }

        /*  It's an error if we get more arguments than we have space for  */
        if (command->argument_count >= MAX_COMMAND_ARGUMENTS) {
            errno = EINVAL;
            return -1;
        }

        command->argument_name[command->argument_count] = tokens[i];
        command->argument_value[command->argument_count] = tokens[i + 1];
        command->argument_count++;

        i += 2;
    }

    return 0;
}
